juju/framework.py

```diff
@@ -573,12 +612,10 @@ def __setattr__(self, key, value):
             raise AttributeError(f"attribute '{key}' cannot be set to {type(value).__name__}: must be int/dict/list/etc")
 
         self._data[key] = _unwrap_stored(self._data, value)
-        self._data.framework.save_snapshot(self._data)
         self.on.changed.emit()
 
-        # TODO Saving a snapshot on every change is not efficient. Instead, the
-        # the framework should offer a pre-commit event that the state can monitor
-        # and save itself at the right time if changes are pending.
+    def is_dirty(self):
```

Comment:
Everything we put on this type clobbers the namespace we offer the user.

Please see the comment above on `__getattr__`.

Reply:
I only use it in `test_basic_state_storage`.

So either `test_basic_state_storage` needs to use `obj.state._data.is_dirty` or there needs to be a "public" method exposed for retrieving this. I can go for the former as we do not use it for anything else yet.

Reply:
Actually, there is another problem since `__getattr__` is reimplemented for BoundStoredState. I cannot just check `obj.state._data.dirty` in the unit test because it will trigger `AttributeError(f"attribute '{key}' is not stored")`.